angr/sim_state.py

```diff
@@ -52,8 +52,9 @@ class SimState(PluginHub):
     :ivar str unicorn:      Control of the Unicorn Engine
     """
 
-    def __init__(self, project=None, arch=None, plugins=None, memory_backer=None, permissions_backer=None, mode=None, options=None,
-                 add_options=None, remove_options=None, special_memory_filler=None, os_name=None, plugin_preset='default', **kwargs):
+    def __init__(self, project=None, arch=None, plugins=None, memory_backer=None, permissions_backer=None, mode=None,
+                 options=None, add_options=None, remove_options=None, special_memory_filler=None, os_name=None,
+                 heap=None, plugin_preset='default', **kwargs):
```

Comment:
do we want to have heap broken out into its own kwarg or to have it be part of the plugins? We can have `heap=xx` on the SimOS state factory level.

Reply:
I think that would be reasonable. The only caveat is that, because of the dependency of the heap plugin on the memory plugin, doing so will add some minor complexity to `SimState`'s `__init__`. Something like this.

```python
        if plugins is not None:
            for n,p in plugins.items():
                if n == 'heap':
                    continue
                self.register_plugin(n, p, inhibit_init=True)
            for p in plugins.values():
                if isinstance(p, angr.state_plugins.heap.SimHeapBase):
                    continue
                p.init_state()
```

And then at the end of `__init__`, we add the following.

```python
        if plugins is not None:
            h = plugins['heap']
            if h is not None:
                self.register_plugin('heap', h)
```

That's not too bad though, and if it is, maybe there's a more elegant way of doing it that I'm ignorant of at the moment. Let me know what you think and I can add the above if this is the direction you want to go in.

Reply:
So! We have a very janky tool to deal with dependencies between plugins - this is the difference between the `set_state` and `init_state` methods. `set_state` is where you need to do any initialization that doesn't depend on other plugins, and then `init_state` is where you depend on other plugins. This doesn't work if you have any code in `init_state` which depends on another plugin's `init_state` having already run, but the memory plugin has no `init_state`, so this shouldn't be a problem at all!

Reply:
So this has brought me back to some problems I had previously while trying to get the plugin initialized at the right time. If I understand things correctly, the `memory` plugin [doesn't actually exist when the rest of the plugins are initialized](https://github.com/angr/angr/blob/master/angr/sim_state.py#L91)--`sym_memory`, `fast_memory`, and `abs_memory` do, but from what I can tell these would be both dangerous to use directly and kind of messy. `SimState` goes on to create the `memory` plugin afterwards depending on what options have been specified, and in my usage I have not seen the args to it be anything other than `None`. I could totally have missed the channel by which the backing memory for `sym_memory` gets passed on to `memory`, but in my experimentation it seems like I end up with differing views of memory depending on which I use, and when. As it is now, I cannot refer to the state's `memory` plugin from the heap plugin's `init_state` without getting an attribute error.

If I move the plugin initialization to be after the `memory` plugin is set up, things work fine, except for the `SimState`'s `_global_condition` attribute [not having been set yet](https://github.com/angr/angr/blob/master/angr/sim_state.py#L146). If I move that up as well, to be set before the plugins are initialized, then things appear to work as they should.

That being said, I feel like I'm playing with fire since `SimState`'s `__init__` is pretty important and attaches to things I'm not too familiar with. Can you verify whether I'm off track here?

Reply:
Forgot to add: the nonexistent `_global_condition` gets referenced by the `memory` plugin when stores are requested during a heap's `init_state`.